sklearn/inspection/_partial_dependence.py

```diff
@@ -592,13 +593,20 @@ def plot_partial_dependence(estimator, X, features, feature_names=None,
         # regression and binary classification
         target_idx = 0
 
-    X = check_array(X)
+    # Use check_array only on lists and other non-array-likes / sparse. Do not
+    # convert DataFrame into a NumPy array.
+    if not(hasattr(X, '__array__') or sparse.issparse(X)):
+        X = check_array(X, force_all_finite='allow-nan', dtype=np.object)
     n_features = X.shape[1]
 
     # convert feature_names to list
     if feature_names is None:
-        # if feature_names is None, use feature indices as name
-        feature_names = [str(i) for i in range(n_features)]
+        if hasattr(X, "loc"):
+            # get the column names for a pandas dataframe
+            feature_names = X.columns.tolist()
```

Comment:
should this not explicitly check for `columns` instead?

Reply:
Until now, we always ducktyped dataframe in this way.

Reply:
Maybe it's time to  have a is_dataframe helper

Reply:
we can leave it for another PR though. I'm happy as is for this PR